bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java

```diff
@@ -0,0 +1,187 @@
+        // Create the SSL context.
+        try {
+            ctx = SSLContext.getInstance("TLS");
+
+            KeyManager[] keyManagers = null;
+            if (conf.isClientSSLAuthentication()) {
+                KeyStore ks = KeyStore.getInstance("pkcs12");
+                String sslKeyStore = conf.getClientSSLKeyStore();
+                InputStream certStream = getClass().getResourceAsStream(sslKeyStore);
+                if (certStream == null) {
+                    LOG.info("Loading BookKeeper client SSLKeyStore from disk path {}", sslKeyStore);
+                    certStream = new FileInputStream(sslKeyStore);
+                } else {
+                    LOG.info("Found BookKeeper client from classpath {}", sslKeyStore);
+                }
+                try {
+                    ks.load(certStream, conf.getClientSSLKeyStorePassword().toCharArray());
+                } finally {
+                    certStream.close();
+                }
+
+                KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
+                kmf.init(ks, conf.getClientSSLKeyStorePassword().toCharArray());
+                keyManagers = kmf.getKeyManagers();
+            }
+
+            if (conf.getVerifySSLCertificates()) {
+                ctx.init(keyManagers, null, null);
+            } else {
+                ctx.init(keyManagers, getTrustManagers(), null);
+            }
+        } catch (KeyStoreException kse) {
+            throw new RuntimeException("Standard keystore type missing", kse);
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new SSLConfigurationException("Standard algorithm missing", nsae);
+        } catch (KeyManagementException | IOException kme) {
+            throw new SSLConfigurationException("Error initializing SSLContext", kme);
+        } catch (CertificateException ce) {
+            throw new SSLConfigurationException("Unable to load keystore", ce);
+        } catch (UnrecoverableKeyException uke) {
+            throw new SSLConfigurationException("Unable to load key manager, possibly wrong password given", uke);
+        }
+    }
+
+    public SSLContext getContext() {
+        return ctx;
+    }
+
+    public SSLEngine getEngine() {
+        SSLEngine engine = ctx.createSSLEngine();
+        engine.setUseClientMode(isClient);
+        if (!isClient) {
+            // allow clients to authenticate
+            engine.setWantClientAuth(true);
+        }
+        if (enabledProtocols != null && !enabledProtocols.isEmpty()) {
+            String[] protocols = enabledProtocols.split(",");
+            engine.setEnabledProtocols(protocols);
+        }
+        if (enabledCipherSuites != null && !enabledCipherSuites.isEmpty()) {
+            String[] ciphers = enabledCipherSuites.split(",");
+            engine.setEnabledCipherSuites(ciphers);
+        }
+        return engine;
+    }
+
+    private TrustManager[] getTrustManagers() {
```

Comment:
IMHO I think that check be done in the AuthProvider.

But we can take care of this directly here.
What do you think ?